// satellite.js/dist/wasm/calculators/ecf-velocity-calculator.js
const DIMENSIONS = 3;
/**
 * Calculator for ECF (Earth-Centered Fixed) velocity.
 *
 * Depends on:
 * @see EciBaseCalculator
 * @see GmstCalculator
 *
 * Raw outputs are always sorted by satellite index first, then by date index, and packed as:
 *   - `Float64Array`, packed as [vx0, vy0, vz0, vx1, vy1, vz1, ...] for each satellite/date pair
 *
 * Provides formatted output under `ecfVelocity` property.
 */
export class EcfVelocityCalculator {
    name = 'ecfVelocity';
    dependencies = ['eci', 'gmst'];
    satellitesCount;
    datesCount;
    module;
    outputPointer;
    init(module, outputPointer, satellitesCount, datesCount) {
        this.module = module;
        this.outputPointer = outputPointer;
        this.satellitesCount = satellitesCount;
        this.datesCount = datesCount;
    }
    getFormattedOutput(satelliteIndex, dateIndex) {
        const rawOutput = this.getRawOutput();
        const index = (satelliteIndex * this.datesCount + dateIndex) * DIMENSIONS;
        return {
            x: rawOutput[index],
            y: rawOutput[index + 1],
            z: rawOutput[index + 2],
        };
    }
    getOutputBufferSize(satellitesCount, datesCount) {
        return satellitesCount * datesCount * DIMENSIONS * Float64Array.BYTES_PER_ELEMENT;
    }
    getRawOutput() {
        return new Float64Array(this.module.HEAP8.buffer, this.outputPointer, this.satellitesCount * this.datesCount * DIMENSIONS);
    }
    getExecutionDescriptor() {
        return {
            ecfVelocityEnabled: true,
            ecfVelocities: this.outputPointer,
        };
    }
}
